File: num_meth/lab_3/gauss.m
global ORDER;
ORDER=3;
A = [1.80 2.50 4.60; 
     3.10 2.30 -1.20; 
     4.51 -1.80 3.60];
b = [2.20 3.60 -1.70];

function solve = gauss(A, b, order)
    tmp = 0;
    for i=1:order
        for j=1:order
            if i != j && A(i,i) == 0 && A(j,i) != 0 && A(i,j) != 0
                for k=1:order
                    A([j,i],:) = A([i,j],:);
                end
                b([i,j]) = b([j,i]);
                break;
            end
        end 
    end
    for i=1:order
        if A(i,i) == 0
            solve = "e";
            return;
        end
        for j=i+1:order
            tmp = A(j,i) / A(i,i);
            for t=i:order
                A(j,t) = A(j,t) - tmp * A(i,t);
            end
            b(j) -= tmp * b(i);
        end
    end
    solve = zeros(order,1);
    for i=order:-1:1
        sum_known_elems = 0;
        for j=1:order
            sum_known_elems = sum_known_elems + A(i,j) * solve(j);
        end
        solve(i) = (b(i) - sum_known_elems) / A(i,i);
    end
end

function resid_res = resid(A, b, solve)
    resid_res = (A * solve).' - b;
end

solve = gauss(A,b,ORDER);
residial = resid(A, b, solve);

if solve == "e"
    printf("Решений нет!\n");
else
    for i=1:ORDER
        printf("X%d = %f\n", i, solve(i));
    end
    printf("=================================\n");
    for i=1:ORDER
        printf("F%d = %f\n", i, residial(i));
    end
    printf("Residial norm: %f\n", max(abs(residial)));
end

File: num_meth/sem6/lab2/utils/gauss.c
#include <stdio.h>
#include <math.h>

#include "gauss.h"

static int prepare_main_elem(double **matrix, double *extension, int order);

int gauss(double **matrix, double *extension, int order, double *solve)
{
	int ret = 0;
	int i,j,k;
	if (matrix == NULL || extension == NULL || solve == NULL) {
		ret = -1;
		goto out;
	}
	
	if ((ret = prepare_main_elem(matrix, extension, order)) != 0) {
		goto out;
	}
		
	for (i = 0; i < order; ++i){
		if (matrix[i][i] == 0){				
			ret = -2;
			goto out;
		}
		for(j = i + 1; j < order; ++j){
			double tmp = matrix[j][i] / matrix[i][i];
			for(k = i; k < order; ++k){
				matrix[j][k] -= tmp * matrix[i][k];
			}
			extension[j] -= tmp * extension[i];
		}
	}
	for (i = 0; i < order; ++i) {
		solve[i] = 0;
	}
	for(i = order - 1; i >= 0; --i){
		double sum_known_elems = 0;
		for(j = i; j < order; ++j){
			sum_known_elems += matrix[i][j] * solve[j];
		}
		solve[i] = (extension[i] - sum_known_elems) / matrix[i][i];
	}

out:	
	return ret;
}

int resid(double **matrix, double *extension, double *solve, int order, double *resid)
{
	int i,j;
	if (matrix == NULL ||
	    extension == NULL ||
	    solve == NULL ||
	    resid == NULL)
	{
		printf("Error (resid): NULL pointer passed\n");
		return -1;		
	}
	for (i = 0; i < order; ++i) {
		int res = 0;
		for (j = 0; j < order; ++j) {
			resid[i] += matrix[i][j] * solve[j];
		}
		resid[i] -= extension[i];
	}
	return 0;
}

int resid_norm(double *resid, int order, double *norm)
{
	int i;
	double max_abs = 0;
	if (resid == NULL || norm == NULL) {
		printf("Error (resid): NULL pointer passed\n");
		return -1;				
	}
	for (i = 0; i < order; ++i) {
		if (fabs(resid[i]) > max_abs) {
			max_abs = fabs(resid[i]);
		}
	}
	*norm = max_abs;
	return 0;
}


static int prepare_main_elem(double **matrix, double *extension, int order)
{
	int i,j,k;
	double tmp = 0;
	for (i = 0; i < order; ++i) {
		for (j = 0; j < order; ++j) {
			if (matrix[i][j] != 0 &&
			    matrix[j][i] != 0 &&
			    matrix[i][i] == 0 &&
			    i != j)
			{
				for (k = 0; k < order; ++k) {
					tmp = matrix[j][k];
					matrix[j][k] = matrix[i][k];
					matrix[i][k] = tmp;
				}
				tmp = extension[j];
				extension[j] = extension[i];
				extension[i] = tmp;
				break;
			}
		}
	}
	return 0;
}
